File: code/ruby/oauth/kiva.rb
#!/usr/local/bin/ruby

require 'rubygems'
require 'oauth'
require 'json'
require 'net/http'
require 'net/https'

auth={}
auth["consumer_key"]    = ''
auth["consumer_secret"] = ''

event_url = "https://dev-vm-01.kiva.org/sites/kiva_api/oauth/request_token.json"
path = "/sites/kiva_api/oauth/request_token.json"
site = "https://dev-vm-01.kiva.org"
callback_url = 'oob'

#==============================
#=Set up signed request to get request token
consumer = OAuth::Consumer.new(
  auth["consumer_key"], 
  auth["consumer_secret"], {
    :oauth_version => "1.0",  
    :site   => site,
    :scheme => :body,
    :http_method => :post
})
req  = consumer.create_signed_request(:post, path)
data = URI.unescape(req.body)
data = Hash[*req.body.split(/=|&/)]

uri = URI.parse(event_url)
http = Net::HTTP.new(uri.host, uri.port)
http.use_ssl = true
http.verify_mode = OpenSSL::SSL::VERIFY_NONE 
request = Net::HTTP::Post.new(path)
request.set_form_data(data)
resp = http.request(request)
token_n_secret = JSON.parse resp.body


#===============================
#=Setup Access token
request_token_consumer = OAuth::Consumer.new(
     auth["consumer_key"], 
     auth["consumer_secret"],
    :oauth_version => "1.0",  
    :site   => site,
    :scheme => :body,
    :http_method => :post)

oauth_url = 'https://dev-vm-01.kiva.org/httpdocs/oauth/authorize?response_type=code&client_id='+auth["consumer_key"]+'&type=web_server'+'&oauth_token='+token_n_secret['oauth_token']+'&oauth_callback='+URI.escape(callback_url)

puts oauth_url
puts "\nEnter the Authorization Code:"
auth_code_from_browser = gets.chomp
ouath_hash_verifier = {:oauth_verifier => auth_code_from_browser.to_s}

t = OAuth::Token.new( token_n_secret['oauth_token'], token_n_secret['oauth_token_secret'])

req  = request_token_consumer.create_signed_request(:post, "/sites/kiva_api/oauth/access_token.json", t, ouath_hash_verifier)
data = URI.unescape(req.body)
data = Hash[*req.body.split(/=|&/)]
data.merge(ouath_hash_verifier)

uri = URI.parse(event_url)
http = Net::HTTP.new(uri.host, uri.port)
http.use_ssl = true
http.verify_mode = OpenSSL::SSL::VERIFY_NONE 
request = Net::HTTP::Post.new("/sites/kiva_api/oauth/access_token.json")
request.set_form_data(data)
resp = http.request(request)

#=====access token
access_token_consumer = OAuth::Consumer.new(
     auth["consumer_key"], 
     auth["consumer_secret"],
    :oauth_version => "1.0",  
    :site   => site,
    :scheme => :body,
    :http_method => :post)

access_token_obj = JSON.parse resp.body
puts access_token_obj.inspect

=begin
at = OAuth::Token.new(access_token_obj['oauth_token'], access_token_obj['oauth_token_secret'])
req  = access_token_consumer.create_signed_request(:post, "/sites/kiva_api/v1/my/account.json", at)
data = URI.unescape(req.body)
data = Hash[*req.body.split(/=|&/)]

uri = URI.parse(event_url)
http = Net::HTTP.new(uri.host, uri.port)
http.use_ssl = true
http.verify_mode = OpenSSL::SSL::VERIFY_NONE 
request = Net::HTTP::Post.new( "/sites/kiva_api/v1/my/account.json")
request.set_form_data(moredata)
resp = http.request(request)
puts resp.inspect
pp resp.inspect
=end

File: code/js/oauth/kiva.js
$(document).ready(function() {

	(function ($, global, undefined) {
		'use strict';

		// You need to set these
		var Consumer = {
			'key': '',
			'secret': '',
			// If you have set up a callback URL at build.kiva.org, you should enter it below. Otherwise leave it as 'oob'
			// Make sure it is URL encoded here
			'callbackUrl': encodeURIComponent('oob')
		};
	
		// This is the URL of the protected resource you want to access
		var resourceUrl = 'https://api.kivaws.org/v1/my/account.json';

		// These should stay the same, probably
		var requestTokenUrl = 'https://api.kivaws.org/oauth/request_token.json';
		var authorizeUrl = 'https://www.kiva.org/oauth/authorize?response_type=code&client_id='+Consumer.key+'&type=web_server&scope=access&oauth_callback='+Consumer.callbackUrl;
		var accessTokenUrl = 'https://api.kivaws.org/oauth/access_token.json';

		// Leave everything below this line alone
	
		// Helper functions
		function generateNonce() {
			return Math.random()*10000000;
		}

		function generateTimestamp() {
			var dateTime = new Date();
			var time = dateTime.getTime();
			return parseInt(time/1000, 10);
		}
	
		function generateSignature(httpMethod, url, nonce, timestamp, token, tokenSecret, verifier) {

			if (tokenSecret === undefined) {
				tokenSecret = '';
			}

			var base = httpMethod.toUpperCase() + "&" + encodeURIComponent(url).toString() + "&"
				+ encodeURIComponent("oauth_callback") + "%3D" + encodeURIComponent(Consumer.callbackUrl)
				+ "%26"
				+ encodeURIComponent("oauth_consumer_key") + "%3D" + encodeURIComponent(Consumer.key)
				+ "%26"
				+ encodeURIComponent("oauth_nonce") + "%3D" + encodeURIComponent(nonce)
				+ "%26"
				+ encodeURIComponent("oauth_signature_method") + "%3D" + encodeURIComponent("HMAC-SHA1")
				+ "%26"
				+ encodeURIComponent("oauth_timestamp") + "%3D" + encodeURIComponent(timestamp);

			if (token !== undefined) {
				base += "%26"
					+ encodeURIComponent("oauth_token") + "%3D" + encodeURIComponent(token);
			}

			if (verifier !== undefined) {
				base += "%26"
					+ encodeURIComponent("oauth_verifier") + "%3D" + encodeURIComponent(verifier);
			}

			base += "%26"
				+ encodeURIComponent("oauth_version") + "%3D" + encodeURIComponent("1.0");

			return global.b64_hmac_sha1(Consumer.secret+'&'+tokenSecret, base)+'=';
		}
	
		function generateHeader(nonce, timestamp, signature, token, verifier) {
			var header = 'OAuth oauth_nonce="'+nonce+'",oauth_callback="'+Consumer.callbackUrl+'",oauth_signature_method="HMAC-SHA1",oauth_timestamp="'+timestamp+'",oauth_consumer_key="'+Consumer.key+'",oauth_signature="'+signature+'"';

			if (token !== undefined) {
				header += ', oauth_token="'+token+'"';
			}

			if (verifier !== undefined) {
				header += ', oauth_verifier="'+verifier+'"';
			}

			header += ', oauth_version="1.0"';

			return header;
		}
	
		function fetchRequestToken(callback) {

			var nonce = generateNonce();
			var timestamp = generateTimestamp();
			var signature = generateSignature('POST', requestTokenUrl, nonce, timestamp);
			var header = generateHeader(nonce, timestamp, signature);

			$.ajax({
				url: requestTokenUrl,
				type: "POST",
				beforeSend: function (request)
				{
					request.setRequestHeader("Authorization", header);
				},
				dataType: "json",
				success: callback,
				error: function (data) {
					global.alert("Error fetching request token");
				}
			});
		}
	
		function fetchAccessToken(requestToken, requestTokenSecret, verifier, callback) {

			var nonce = generateNonce();
			var timestamp = generateTimestamp();
			var signature = generateSignature('POST', accessTokenUrl, nonce, timestamp, encodeURIComponent(requestToken), encodeURIComponent(requestTokenSecret), encodeURIComponent(verifier));
			var header = generateHeader(nonce, timestamp, signature, requestToken, verifier);

			$.ajax({
				url: accessTokenUrl,
				type: "POST",
				beforeSend: function (request)
				{
					request.setRequestHeader("Authorization", header);
				},
				dataType: "json",
				success: callback,
				error: function (data) {
					global.alert("Error fetching access token");
				}
			});
		}
	
		function fetchResource(url, accessToken, accessTokenSecret, callback) {

			var nonce = generateNonce();
			var timestamp = generateTimestamp();
			var signature = generateSignature('GET', url, nonce, timestamp, encodeURIComponent(accessToken), encodeURIComponent(accessTokenSecret));
			var header = generateHeader(nonce, timestamp, signature, accessToken);

			$.ajax({
				url: url,
				type: "GET",
				beforeSend: function (request)
				{
					request.setRequestHeader("Authorization", header);
				},
				dataType: "json",
				success: callback,
				error: function (data) {
					global.alert("Error fetching resource");
				}
			});
		}

		function init() {
			// This is the function that pulls it all together, through a series off callbacks triggered at every step
			$('body').append('<button id="authBtn">Authorize</button>');
			$('#authBtn').click(function() {
				fetchRequestToken(function(data) {
					global.alert('Redirecting to the authorize page. Copy the authorization code and paste it back on this page.');
					global.open(authorizeUrl+'&oauth_token='+data.oauth_token, '_blank');
					global.focus();

					$('#authBtn').remove();
					$('body').append('<input id="oauth_verifier" type="text"/>').append('<button id="accessBtn">Get data</button>');
					$('#accessBtn').click(function() {
						fetchAccessToken(data.oauth_token, data.oauth_token_secret, $('#oauth_verifier').val(), function(data) {
							fetchResource(resourceUrl, data.oauth_token, data.oauth_token_secret, function(data) {
								global.alert('Hello '+data.user_account.first_name+' '+data.user_account.last_name);
							});
						});
					});
				});
			});
		}

		init();

	}(jQuery, window));
});
